File: Examenes-EDD-3/2021-Septiembre/Haskell/SparseMatrix.hs
module SparseMatrix (
    SparseMatrix,
    sparseMatrix,
    set,
    get,
    add,
    transpose,
    fromList
) where

import qualified DataStructures.Dictionary.AVLDictionary as D -- Asumiendo que tienes un módulo de diccionario AVL
import Data.Maybe (fromMaybe)

-- Representación de un índice de matriz
data Index = Idx Int Int deriving (Eq, Ord, Show)

-- Tipo de datos para representar una matriz dispersa
data SparseMatrix = SM Int Int (D.Dictionary Index Int)

-- Construye una matriz nula de dimensiones dadas
sparseMatrix :: Int -> Int -> SparseMatrix
sparseMatrix fil col = SM fil col D.empty

-- Función privada: obtiene el valor en un índice dado
value :: SparseMatrix -> Index -> Int
value (SM _ _ dict) idx = fromMaybe 0 (D.valueOf idx dict)

-- Función privada: actualiza un valor dado un índice
update :: SparseMatrix -> Index -> Int -> SparseMatrix
update (SM fil col dict) idx val
    | val == 0  = SM fil col (D.delete idx dict)
    | otherwise = SM fil col (D.updateOrInsert idx (const val) val dict)

-- Función privada: construye un índice con validación
index :: SparseMatrix -> Int -> Int -> Index
index (SM filMax colMax _) fil col 
    | fil < 0 || col < 0 || fil > filMax || col > colMax = error "index out of bounds"
    | otherwise                                          = Idx fil col

-- Función pública: actualiza el valor en una posición (valida el índice)
set :: SparseMatrix -> Int -> Int -> Int -> SparseMatrix
set sm fil col val = update sm (index sm fil col) val

-- Función pública: obtiene el valor de una posición (valida el índice)
get :: SparseMatrix -> Int -> Int -> Int
get sm fil col = value sm (index sm fil col)

-- Suma de dos matrices dispersas
add :: SparseMatrix -> SparseMatrix -> SparseMatrix
add sm1@(SM fil1 col1 dict1) sm2@(SM fil2 col2 dict2)
    | fil1 /= fil2 || col1 /= col2 = error "matrices have not the same dimension"
    | otherwise                    = SM fil1 col1 (addRec (D.keysValues dict2))
        where 
            addRec :: [(Index, Int)] -> D.Dictionary Index Int
            addRec [] = dict1
            addRec ((idx, v):xs) = 
                case D.valueOf idx dict1 of 
                    Nothing -> D.insert idx v (addRec xs)
                    Just v' -> if v + v' == 0 then D.delete idx (addRec xs) else D.updateOrInsert idx (const (v + v')) (v + v') (addRec xs)

-- Traspuesta de una matriz dispersa
transpose :: SparseMatrix -> SparseMatrix
transpose (SM fil col dict) = SM col fil (foldr (\(Idx f c, v) -> D.insert (Idx c f) v) D.empty (D.keysValues dict))

-- Crea una matriz dispersa a partir de una lista [fila, col, val, fila, col, val, ...]
fromList :: Int -> Int -> [Int] -> SparseMatrix
fromList fil col xs
    | length xs `mod` 3 /= 0 = error "list's length is not multiple of 3"
    | otherwise              = SM fil col (fromListRec xs)
        where 
            fromListRec :: [Int] -> D.Dictionary Index Int 
            fromListRec []          = D.empty
            fromListRec (f:c:v:xs)  = D.insert (Idx f c) v (fromListRec xs)
-- Complejidad esperada: O(n log k), donde n = longitud de la lista / 3, k = elementos distintos insertados.

File: Examenes-EDD-4/2021-Septiembre/Java/sparseMatrix/SparseMatrix.java
package sparseMatrix;

import java.util.List;
import java.util.TreeMap;

public class SparseMatrix {
    private final int rows;
    private final int cols;
    private final TreeMap<Index, Integer> values;

    public SparseMatrix(int rows, int cols) {
        this.rows = rows;
        this.cols = cols;
        this.values = new TreeMap<>();
    }

    // Devuelve el valor en una posición (privado)
    private int value(Index idx) {
        return values.containsKey(idx) ? values.get(idx) : 0;
    }

    // Modifica el valor en una posición (privado)
    private void update(Index idx, int val) {
        if(val == 0)
            values.remove(idx);
        else 
            values.put(idx, val);
    }

    // Devuelve un índice validado
    private Index index(int r, int c) {
        if(r < 0 || c < 0 || r > rows || c > cols)
            throw new IllegalArgumentException("index out of bounds");

        return new Index(r, c);
    }

    // Pública: set
    public void set(int r, int c, int val) {
        if(r < 0 || c < 0 || r > rows || c > cols)
            throw new IllegalArgumentException("index out of bounds");

        update(index(r, c), val);
    }

    // Pública: get
    public int get(int r, int c) {
        if(r < 0 || c < 0 || r > rows || c > cols)
            throw new IllegalArgumentException("index out of bounds");

        return value(index(r, c));
    }

    // Pública: suma de dos matrices
    public SparseMatrix add(SparseMatrix other) {
        if(rows != other.rows || cols != other.cols)
            throw new IllegalArgumentException("the given matrices have not the same dimension");

        SparseMatrix sp = new SparseMatrix(rows, cols);
        
        for(Index idx : values.keySet())
            sp.values.put(idx, values.get(idx));

        for(Index idx : other.values.keySet())
            sp.update(idx, sp.value(idx) + other.value(idx));

        return sp;
    }

    // Pública: traspuesta
    public SparseMatrix transpose() {
        SparseMatrix sp = new SparseMatrix(cols, rows);

        for(Index idx :values.keySet())
            sp.set(idx.getColumn(), idx.getRow(), values.get(idx));

        return sp;
    }

    // Pública: fromList
    public static SparseMatrix fromList(int rows, int cols, List<Integer> data) {
        if(rows < 0 || cols < 0)
            throw new IllegalArgumentException("index out of bounds");

        if(data.size() % 3 != 0)
            throw new IllegalArgumentException("the length of the given list is not multiple of 3");

        SparseMatrix sp = new SparseMatrix(rows, cols);

        for(int i = 0; i < data.size(); i += 3)
            sp.set(data.get(i), data.get(i + 1), data.get(i + 2));

        return sp;
    }
}
